network/src/primitives.rs

```diff
@@ -0,0 +1,39 @@
+use super::*;
+use async_trait::async_trait;
+use tokio::sync::mpsc;
+
+#[async_trait]
+pub trait PeerDiscoveryPrimitive {
+    /// Remains online on the network indefinitely,
+    /// responding to discovery requests from other nodes,
+    /// updating `known_peers`.
+    async fn serve(
+        network_config: &NetworkConfig,
+        initially_known_peers: Vec<Peer>,
+        known_peers_send: mpsc::Sender<Peer>,
```

Comment:
Why does it need `known_peers_send` though it will return `SharedKnownPeers`?

Reply:
It is not needed. Fixed.

Reply:
ok